File: gtci/16-subsets/cpp/9-count-of-structurally-unique-bsts.cpp
#include <cassert>
#include <iostream>
#include <unordered_map>

class Solution {
public:
  int countTrees(int n) {
    if (memo.contains(n))
      return memo[n];

    if (n <= 1)
      return 1;
    
    int count = 0;

    for (int rootVal = 1; rootVal < n + 1; rootVal++) {
      int leftSubtreeCount = countTrees(rootVal - 1);
      int rightSubtreeCount = countTrees(n - rootVal);

      count += leftSubtreeCount * rightSubtreeCount;
    }

    memo[n] = count;
    return count;
  }
private:
  std::unordered_map<int, int> memo;
};

int main() {
  Solution s;

  assert(s.countTrees(2) == 2);
  assert(s.countTrees(3) == 5);

  std::cout << "All test cases passed." << std::endl;
  return 0;
}

File: gtci/16-subsets/python3/9-count-of-structurally-unique-bsts.py
class Solution:
    def __init__(self):
        self.memo = {}

    def countTrees(self, n):
        if n in self.memo:
            return self.memo[n]

        if n <= 1:
            return 1

        count = 0

        for root_val in range(1, n + 1):
            left_subtree_count = self.countTrees(root_val - 1)
            right_subtree_count = self.countTrees(n - root_val)

            count += left_subtree_count * right_subtree_count

        self.memo[n] = count
        return count


if __name__ == "__main__":
    sol = Solution()

    assert sol.countTrees(2) == 2
    assert sol.countTrees(3) == 5

    print("All test cases passed.")
